ethexe/runtime/common/src/schedule.rs

```diff
@@ -0,0 +1,86 @@
+use crate::{
+    state::{ComplexStorage, ProgramState, Storage},
+    InBlockTransitions,
+};
+use anyhow::Result;
+use gear_core::{ids::ProgramId, tasks::TaskHandler};
+use gprimitives::{ActorId, CodeId, MessageId, ReservationId, H256};
+
+pub struct Handler<'a, S: Storage> {
+    pub in_block_transitions: &'a mut InBlockTransitions,
+    pub storage: &'a S,
+}
+
+impl<S: Storage> Handler<'_, S> {
+    pub fn update_state(
+        &mut self,
+        program_id: ProgramId,
+        f: impl FnOnce(&mut ProgramState) -> Result<()>,
+    ) -> H256 {
+        crate::update_state(self.in_block_transitions, self.storage, program_id, f)
+    }
+
+    pub fn update_state_with_storage(
+        &mut self,
+        program_id: ProgramId,
+        f: impl FnOnce(&S, &mut ProgramState) -> Result<()>,
+    ) -> H256 {
+        crate::update_state_with_storage(self.in_block_transitions, self.storage, program_id, f)
+    }
+}
+
+impl<'a, S: Storage> TaskHandler<ActorId> for Handler<'a, S> {
+    fn remove_from_mailbox(&mut self, _user_id: ActorId, _message_id: MessageId) -> u64 {
+        unimplemented!("TODO (breathx)")
+    }
+    fn send_dispatch(&mut self, _stashed_message_id: MessageId) -> u64 {
+        unimplemented!("TODO (breathx)")
+    }
+    fn send_user_message(&mut self, _stashed_message_id: MessageId, _to_mailbox: bool) -> u64 {
+        unimplemented!("TODO (breathx)")
+    }
+    // TODO (breathx): consider deprecation of delayed wakes + non-concrete waits.
+    fn wake_message(&mut self, program_id: ProgramId, message_id: MessageId) -> u64 {
+        log::trace!("Running scheduled task wake message {message_id} to {program_id}");
+
+        // TODO (breathx): don't update state if not changed?
+        self.update_state_with_storage(program_id, |storage, state| {
+            let Some(((dispatch, _expiry), new_waitlist_hash)) = storage
```

Comment:
Maybe unreachable for case when `_expiry < current`.

Reply:
unfortunately, I think it could be reached in some cases. we need to support outdated tasks and handle cases why we've missed something